lib/staging/tls/tls.cpp

```diff
@@ -0,0 +1,1459 @@
+
+            OCSP_RESPONSE* resp{nullptr};
+
+            if (filename != nullptr) {
+                resp = load_ocsp(filename);
+                if (resp == nullptr) {
+                    bResult = false;
+                }
+            }
+
+            if (resp != nullptr) {
+                updates[digest] = std::shared_ptr<OCSP_RESPONSE>(resp, &::OCSP_RESPONSE_free);
+            }
+        }
+
+        {
+            std::lock_guard lock(mux);
+            m_context->cache.swap(updates);
+        }
+    }
+
+    return bResult;
+}
+
+std::shared_ptr<OcspResponse> OcspCache::lookup(const openssl::sha_256_digest_t& digest) {
+    assert(m_context != nullptr);
+
+    std::shared_ptr<OcspResponse> resp;
+    std::lock_guard lock(mux);
+    if (const auto itt = m_context->cache.find(digest); itt != m_context->cache.end()) {
+        resp = itt->second;
+    } else {
+        log_error("OcspCache::lookup: not in cache: " + to_string(digest));
+    }
+
+    return resp;
+}
+
+bool OcspCache::digest(openssl::sha_256_digest_t& digest, const x509_st* cert) {
+    assert(cert != nullptr);
+
+    bool bResult{false};
+    const ASN1_BIT_STRING* signature{nullptr};
+    const X509_ALGOR* alg{nullptr};
+    X509_get0_signature(&signature, &alg, cert);
+    if (signature != nullptr) {
+        unsigned char* data{nullptr};
+        const auto len = i2d_ASN1_BIT_STRING(signature, &data);
+        if (len > 0) {
+            bResult = openssl::sha_256(data, len, digest);
+        }
+        OPENSSL_free(data);
+    }
+
+    return bResult;
+}
+
+// ----------------------------------------------------------------------------
+// CertificateStatusRequestV2
+
+bool CertificateStatusRequestV2::set_ocsp_response(const openssl::sha_256_digest_t& digest, SSL* ctx) {
+    bool bResult{false};
+    auto response = m_cache.lookup(digest);
+    if (response) {
+        unsigned char* der{nullptr};
+        auto len = i2d_OCSP_RESPONSE(response.get(), &der);
+        if (len > 0) {
+            bResult = SSL_set_tlsext_status_ocsp_resp(ctx, der, len) == 1;
+            if (bResult) {
+                SSL_set_tlsext_status_type(ctx, TLSEXT_STATUSTYPE_ocsp);
+            } else {
+                log_error((std::string("SSL_set_tlsext_status_ocsp_resp")));
+                OPENSSL_free(der);
+            }
+        }
+    }
+    return bResult;
+}
+
+int CertificateStatusRequestV2::status_request_cb(SSL* ctx, void* object) {
+    // returns:
+    // - SSL_TLSEXT_ERR_OK response to client via SSL_set_tlsext_status_ocsp_resp
+    // - SSL_TLSEXT_ERR_NOACK no response to client
+    // - SSL_TLSEXT_ERR_ALERT_FATAL abort connection
+    bool bSet{false};
+    int result = SSL_TLSEXT_ERR_NOACK;
+    openssl::sha_256_digest_t digest{};
+
+    if (ctx != nullptr) {
+        const auto* cert = SSL_get_certificate(ctx);
+        const auto* name = X509_get_subject_name(cert);
+        bSet = OcspCache::digest(digest, cert);
+        std::cout << "OCSP:  ";
```

Comment:
std::cout should have been removed. (or been in a #ifdef block)
EVLOG isn't used because it has a memory leak (according to valgrind memcheck).
That makes it much harder to see if there are issues with the code under test.
openssl::log_error() has a #ifdef to prevent EVLOG being used for unit tests. EVLOG is used for the running system.

Reply:
The idea was that those libs are kind of everest independent. An alternative would be to use a logging callback from the lib and the module can then register a callback function that uses EVLOG

Reply:
removed EVLOG. logging now determined by external handler